pipe/terminator.py

```diff
@@ -0,0 +1,140 @@
+import builtins
+import functools
+import pipe
+
+from collections import deque
+
+
+class Terminator:
+    """
+    Terminate a pipeline, performing a final operation or returning a final result.
+
+    Unlike Pipe elements, Terminators do not necessarily return iterables, but rather final results of a pipeline.
+    Terminators can also perform final operations such as `foreach` and `reduce`.
+
+    Represent a termination to a Pipe :
+    Described as :
+        as_list = Terminator(lambda iterable: list(iterable))
+    then the following are equivalent :
+        list(range(10) | is_even)
+        range(10) | is_even > as_list
+
+    Or represent a terminating Function :
+        It's a function returning a Terminator
+    Described as :
+        join = Terminator(lambda iterable, sep: sep.join(iterable))
+    and used as :
+        range(10) | is_even > join("-")
+        # 0-2-4-6-8
+
+    Or represent a terminating operation :
+    Described as :
+        foreach = Terminator(lambda iterable, f: f(x) for x in iterable)
+    and used as :
+        range(10) | is_even > foreach(lambda x: print(x))
+    """
+
+    def __init__(self, function):
+        self.function = function
+        functools.update_wrapper(self, function)
+
+    def __lt__(self, other):
+        return self.function(other)
+
+    def __call__(self, *args, **kwargs):
+        return Terminator(
+            lambda iterable, *args2, **kwargs2: self.function(
+                iterable, *args, *args2, **kwargs, **kwargs2
+            )
+        )
+
+
+#######################
+# Collector Terminators
+#######################
+
+
+def _identity(x):
+    return x
+
+
+@Terminator
+def as_dict(iterable, key_func=_identity, value_func=_identity):
+    if key_func is _identity and value_func is _identity:
+        return dict(iterable)
+
+    return dict(iterable | pipe.map(lambda x: (key_func(x), value_func(x))))
```

Comment:
I think the as_dict should it a Pipe of pairs, instead of duplicating the value, like:

    [(1, 2), (3, 4), (5, 6)] > as_dict == {1: 2, 3: 4, 5: 6}

This change your test from:

```python
    data = ["banana", "grapefruit", "guava"]
    expected = {"BANANA": 6, "GRAPEFRUIT": 10, "GUAVA": 5}
    actual = data > term.as_dict(
        key_func=lambda e: e.upper(),
        value_func=lambda e: len(e),
    )
    assert isinstance(actual, dict)
    assert actual == expected
```

to:

```python
    data = ["banana", "grapefruit", "guava"]
    expected = {"BANANA": 6, "GRAPEFRUIT": 10, "GUAVA": 5}
    actual = data | map(lambda fruit: (fruit.upper(), len(fruit)) > term.as_dict
    assert isinstance(actual, dict)
    assert actual == expected
```

Which I find more natural. It may completly eliminate the need of key_func and value_func?

Reply:
Oh that is entirely in line with how I intended `as_dict` and `as_set` to work :D. I am copying the idea from java streams, where you can provide a key or value function if you want, or if you already have a stream of pairs then you can just call `as_dict` with no arguments. (It's technically not that clean in java, but basically like that)

So the intention was that `key_func` and `value_func` exist as a convenience if writing your terminator like that would be more clear vs using `map`, but you absolutely can just call `as_dict` with no arguments if you have a pipe of pairs or would prefer to use `map`.